src/order/filter/basic.lean

```diff
@@ -552,14 +556,30 @@ lemma eventually.mono {p q : α → Prop} {f : filter α} (hp : ∀ᶠ x in f, p
   ∀ᶠ x in f, q x :=
 hp.mp (f.eventually_of_forall hq)
 
-lemma eventually.congr {f : filter α} {β : Type*} {u v : α → β} {p : β → Prop}
-  (h : ∀ᶠ a in f, u a = v a) (h' : ∀ᶠ a in f, p (u a)) : ∀ᶠ a in f, p (v a) :=
-h'.mp (h.mono (by simp {contextual := tt}))
+@[simp] lemma eventually_and {p q : α → Prop} {f : filter α} :
+  (∀ᶠ x in f, p x ∧ q x) ↔ (∀ᶠ x in f, p x) ∧ (∀ᶠ x in f, q x) :=
+⟨λ h, ⟨h.mono $ λ _, and.left, h.mono $ λ _, and.right⟩, λ h, h.1.and h.2⟩
+
+lemma eventually.congr {f : filter α} {p q : α → Prop} (h' : ∀ᶠ x in f, p x)
+  (h : ∀ᶠ x in f, p x ↔ q x) : ∀ᶠ x in f, q x :=
+h'.mp (h.mono $ λ x hx, hx.mp)
+
+-- `@[congr]` doesn't accept this lemma with `∀ᶠ` notation
+@[congr] lemma eventually_congr {f : filter α} {p q : α → Prop} (h : ∀ᶠ x in f, p x ↔ q x) :
```

Comment:
I'm surprised this works as a congruence lemma.  The `h` assumption doesn't look like the simplifier could handle it  (usually it is `∀ ..., R p q` where `R` is a relation like `↔` or `=`).

Another issue is that `f` is constant.  So you cannot simplify the filter.

I would have expected the following congruence lemma (which is also the default built-in one, so you wouldn't need to mark it `@[congr]`):
```lean
@[congr] lemma eventually_congr {f g : filter α} {p q : α → Prop}
    (hf : f = g) (hp : ∀ x, p x ↔ q x) :
  f.eventually p ↔ g.eventually q :=
sorry
```

The same comment also applies to the two other congruence lemmas for `liminf`.

Reply:
How can I test whether it actually work, or only pretends to work?

Reply:
You can check the proof term or the output with `set_option trace.simplify.congruence true`:
```lean
set_option trace.simplify.congruence true
example {α} (f : filter α) (p q : α → Prop) (h : ∀ x, p x ↔ true) :
    ∀ᶠ x in f, p x :=
by show_term { simp* }
```

Reply:
I removed `@[congr]`, at least for now.